File: dereplication_pilot/R/longitudinal_similarity.R
# R port of longitudinal_similarity.py -- cluster-sharing (Jaccard) between every pair
# of a participant's own single-sample timepoints, vs. real calendar days between them.
# Run from dereplication_pilot/.

source("R/utils.R")
library(lubridate)

METADATA <- "/maps/projects/hansen_ol-AUDIT/scratch/NILU/metagenomes/to_bin_sample_host_date.tsv"

load_sample_dates <- function() {
  read_tsv(METADATA, col_types = cols(.default = "c"), progress = FALSE) %>%
    mutate(group = paste0("spaS", str_remove(Sample, "^S_")),
           participant = paste0("p", Subject),
           date = dmy(Time)) %>%
    filter(!is.na(date)) %>%
    select(group, participant, date)
}

run_tier <- function(tier, dates) {
  out_rows <- list()

  for (participant in all_participants()) {
    p <- file.path(PILOT_DIR, participant, tier, "cluster_membership.tsv")
    if (!file.exists(p)) next
    membership <- load_cluster_membership(participant, tier)

    clusters_of_group <- membership %>%
      filter(!is_coassembly(source_group)) %>%
      group_by(source_group) %>%
      summarise(clusters = list(unique(cluster_id)), .groups = "drop")

    groups <- clusters_of_group$source_group
    if (length(groups) < 2) next

    for (i in seq_along(groups)) {
      for (j in seq_len(length(groups))) {
        if (j <= i) next
        gi <- groups[i]; gj <- groups[j]
        di <- dates$date[dates$group == gi]
        dj <- dates$date[dates$group == gj]
        if (length(di) == 0 || length(dj) == 0) next

        ci <- clusters_of_group$clusters[[i]]
        cj <- clusters_of_group$clusters[[j]]
        shared <- length(intersect(ci, cj))
        uni <- length(union(ci, cj))

        out_rows[[length(out_rows) + 1]] <- tibble(
          participant = participant, group_i = gi, group_j = gj,
          days_between = abs(as.integer(dj[1] - di[1])),
          shared_clusters = shared, n_clusters_i = length(ci), n_clusters_j = length(cj),
          jaccard = if (uni > 0) shared / uni else 0
        )
      }
    }
  }

  bind_rows(out_rows)
}

if (TRUE) {  # always run main logic on source() or Rscript alike -- sys.nframe()==0 broke under nested source()
  dates <- load_sample_dates()
  for (tier in c("orfs", "contigs")) {
    result <- run_tier(tier, dates)
    write_tsv(result, file.path(PILOT_DIR, sprintf("R_longitudinal_similarity_%s.tsv", tier)))
    cat(sprintf("[%s] wrote %d pairs, days %d-%d, jaccard %.3f-%.3f\n",
                tier, nrow(result), min(result$days_between), max(result$days_between),
                min(result$jaccard), max(result$jaccard)))
  }
}

File: dereplication_pilot/longitudinal_similarity.py
#!/usr/bin/env python3
"""Gene-cluster similarity between every pair of a participant's own single-sample
timepoints, vs. real calendar days between them (from to_bin_sample_host_date.tsv).
Tests whether shared-gene-cluster fraction decays with time-distance -- a real
strain-turnover question, and informs whether "annotate once" stays valid over a
long study or needs periodic refresh.

Output: one row per (participant, group_i, group_j) pair, for the Rmd to plot.
"""
import csv
import glob
from collections import defaultdict
from datetime import datetime

METADATA = "/maps/projects/hansen_ol-AUDIT/scratch/NILU/metagenomes/to_bin_sample_host_date.tsv"
PILOT_DIR = "/maps/projects/hansen_ol-AUDIT/scratch/NILU/metagenomes/assembly_annotation_wf/dereplication_pilot"


def load_sample_dates():
    """spaS<n> -> (participant, datetime)"""
    dates = {}
    with open(METADATA) as fh:
        reader = csv.DictReader(fh, delimiter="\t")
        for row in reader:
            n = row["Sample"].replace("S_", "")
            group = f"spaS{n}"
            try:
                d = datetime.strptime(row["Time"], "%d/%m/%Y")
            except ValueError:
                continue
            dates[group] = (row["Subject"], d)
    return dates


def run_tier(tier, dates):
    out_rows = []

    for mfile in sorted(glob.glob(f"{PILOT_DIR}/p*/{tier}/cluster_membership.tsv")):
        participant = mfile.split("/")[-3]
        clusters_of_group = defaultdict(set)  # group -> set of cluster_ids it has a member in
        with open(mfile) as fh:
            reader = csv.DictReader(fh, delimiter="\t")
            for row in reader:
                clusters_of_group[row["source_group"]].add(row["cluster_id"])

        single_groups = sorted(g for g in clusters_of_group if not g.startswith("mh_p"))
        for i in range(len(single_groups)):
            for j in range(i + 1, len(single_groups)):
                gi, gj = single_groups[i], single_groups[j]
                if gi not in dates or gj not in dates:
                    continue
                _, di = dates[gi]
                _, dj = dates[gj]
                days = abs((dj - di).days)

                ci, cj = clusters_of_group[gi], clusters_of_group[gj]
                shared = len(ci & cj)
                union = len(ci | cj)
                jaccard = shared / union if union else 0.0

                out_rows.append((participant, gi, gj, days, shared, len(ci), len(cj), jaccard))

    out_path = f"{PILOT_DIR}/longitudinal_similarity_{tier}.tsv"
    with open(out_path, "w") as fh:
        fh.write("participant\tgroup_i\tgroup_j\tdays_between\tshared_clusters\tn_clusters_i\tn_clusters_j\tjaccard\n")
        for row in out_rows:
            fh.write("\t".join(str(x) for x in row) + "\n")

    print(f"[{tier}] wrote {len(out_rows)} pairs to {out_path}")
    if out_rows:
        days_list = [r[3] for r in out_rows]
        jac_list = [r[7] for r in out_rows]
        print(f"[{tier}] days_between range: {min(days_list)}-{max(days_list)}")
        print(f"[{tier}] jaccard range: {min(jac_list):.3f}-{max(jac_list):.3f}")


def main():
    dates = load_sample_dates()
    for tier in ("orfs", "contigs"):
        run_tier(tier, dates)


if __name__ == "__main__":
    main()
